cli/cmd/upgrade.go

```diff
@@ -0,0 +1,193 @@
+)
+
+func newUpgradeOptionsWithDefaults() *upgradeOptions {
+	return &upgradeOptions{newInstallOptionsWithDefaults()}
+}
+
+func newCmdUpgrade() *cobra.Command {
+	options := newUpgradeOptionsWithDefaults()
+	flags := options.flagSet(pflag.ExitOnError)
+
+	cmd := &cobra.Command{
+		Use:   "upgrade [flags]",
+		Short: "Output Kubernetes configs to upgrade an existing Linkerd control plane",
+		Long:  "Output Kubernetes configs to upgrade an existing Linkerd control plane.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// We need a Kubernetes client to fetch configs and issuer secrets.
+			k, err := options.newK8s()
+			if err != nil {
+				return fmt.Errorf("failed to create a kubernetes client: %s", err)
+			}
+
+			// We fetch the configs directly from kubernetes because we need to be able
+			// to upgrade/reinstall the control plane when the API is not available; and
+			// this also serves as a passive check that we have privileges to access this
+			// control plane.
+			configs, err := fetchConfigs(k)
+			if err != nil {
+				return fmt.Errorf("could not fetch configs from kubernetes: %s", err)
+			}
+
+			// We recorded flags during a prior install. If we haven't overridden the
+			// flag on this upgrade, reset that prior value as if it were specified now.
+			setOptionsFromInstall(flags, configs.GetInstall())
+
+			if err = options.validate(); err != nil {
+				return err
+			}
+
+			// Save off the updated set of flags into the installOptions so it gets
+			// persisted with the upgraded config.
+			options.recordFlags(flags)
+
+			// Update the configs from the synthesized options.
+			options.overrideConfigs(configs, map[string]string{})
+			configs.GetInstall().Flags = options.recordedFlags
+
+			values, err := options.buildValuesWithoutIdentity(configs)
+			if err != nil {
+				return fmt.Errorf("could not build install configuration: %s", err)
+			}
+
+			identityValues, err := fetchIdentityValues(k, options.controllerReplicas, configs.GetGlobal().GetIdentityContext())
+			if err != nil {
+				fmt.Fprintln(os.Stderr, "Unable to fetch the existing issuer credentials from Kubernetes.")
+				fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+				os.Exit(1)
+			}
+			values.Identity = identityValues
+
+			if err = values.render(os.Stdout, configs); err != nil {
+				return fmt.Errorf("could not render install configuration: %s", err)
+			}
+
+			return nil
+		},
+	}
+
+	cmd.PersistentFlags().AddFlagSet(flags)
+	return cmd
+}
+
+func setOptionsFromInstall(flags *pflag.FlagSet, install *pb.Install) {
+	for _, i := range install.GetFlags() {
+		fmt.Printf("flags: setting: %s %s\n", i.GetName(), i.GetValue())
+		if f := flags.Lookup(i.GetName()); f != nil && !f.Changed {
+			f.Value.Set(i.GetValue())
+			f.Changed = true
+		}
+	}
+}
+
+func (options *upgradeOptions) newK8s() (*kubernetes.Clientset, error) {
+	if options.ignoreCluster {
+		return nil, errors.New("--ignore-cluster cannot be used with upgrade")
```

Comment:
We should be able to get rid of the `ignore-cluster` flag for upgrade, which still shows in the help screen. Maybe extracting it from `options.flagSet()` and adding it inside `install.go`'s and `inject.go`'s `newCmdInstall()` ...